biz.aQute.bndlib/src/aQute/bnd/osgi/PermissionGenerator.java

```diff
@@ -0,0 +1,145 @@
+package aQute.bnd.osgi;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import aQute.bnd.header.Attrs;
+import aQute.bnd.osgi.Descriptors.PackageRef;
+
+public class PermissionGenerator {
+	public static final String	KEY					= "permissions";
+
+	public static final String	PARAM_ALL			= "all";
+	public static final String	PARAM_ADMIN			= "admin";
+	public static final String	PARAM_CAPABILITY	= "capability";
+	public static final String	PARAM_PACKAGES		= "packages";
+	public static final String	PARAM_SERVICES		= "services";
+
+	private final Builder		builder;
+	private final List<String>	args;
+	private final StringBuilder	sb;
+
+	public PermissionGenerator(Builder builder, final String... args) {
+		this.builder = builder;
+
+		if (args.length == 2 && PARAM_ALL.equals(args[1])) {
+			this.args = Arrays.asList(PARAM_ADMIN, PARAM_CAPABILITY, PARAM_PACKAGES, PARAM_SERVICES);
+		} else {
+			this.args = Arrays.asList(args);
+		}
+
+		sb = new StringBuilder();
+	}
+
+	public String generate() {
+		sb.setLength(0);
+
+		for (String arg : args) {
+			if (PARAM_ADMIN.equals(arg)) {
+				generateAdmin();
+			} else if (PARAM_CAPABILITY.equals(arg)) {
+				generateCapability();
+			} else if (PARAM_PACKAGES.equals(arg)) {
+				generatePackages();
+			} else if (PARAM_SERVICES.equals(arg)) {
+				generateServices();
+			} else if (KEY.equals(arg))
```

Comment:
Ah. The macro processor must put the macro name in arg[0]. Since the code does a foreach loop on the args array, it needs to ignore element 0 rather then using a for loop with an index starting at 1.